kore/src/Kore/Step/Simplification/Exists.hs

```diff
@@ -156,41 +167,76 @@ simplifyEvaluated variable simplified
   | OrPattern.isTrue simplified  = return simplified
   | OrPattern.isFalse simplified = return simplified
   | otherwise = do
-    evaluated <- traverse (makeEvaluate variable) simplified
+    evaluated <- traverse (makeEvaluate [variable]) simplified
     return (OrPattern.flatten evaluated)
 
-{-| evaluates an 'Exists' given its two 'Pattern' children.
+{-| Evaluates a multiple 'Exists' given a pattern and a list of
+variables which are existentially quantified in the pattern. This
+also sorts the list of variables to ensure that those which are present
+in the substitution are evaluated with the pattern first.
 
 See 'simplify' for detailed documentation.
 -}
 makeEvaluate
-    :: (SimplifierVariable variable, MonadSimplify simplifier)
-    => ElementVariable variable
+    :: forall simplifier variable
+    .  (SimplifierVariable variable, MonadSimplify simplifier)
+    => [ElementVariable variable]
     -> Pattern variable
     -> simplifier (OrPattern variable)
-makeEvaluate variable original
-  = fmap OrPattern.fromPatterns $ Branch.gather $ do
-    normalized <- simplifyCondition original
-    let Conditional { substitution = normalizedSubstitution } = normalized
-    case splitSubstitution variable normalizedSubstitution of
-        (Left boundTerm, freeSubstitution) ->
-            makeEvaluateBoundLeft
-                variable
-                boundTerm
-                normalized { Conditional.substitution = freeSubstitution }
-        (Right boundSubstitution, freeSubstitution) -> do
-            matched <- Monad.Trans.lift $ matchesToVariableSubstitution
-                variable
-                normalized { Conditional.substitution = boundSubstitution }
-            if matched
-                then return normalized
-                    { Conditional.predicate = Predicate.makeTruePredicate_
-                    , Conditional.substitution = freeSubstitution
-                    }
-                else makeEvaluateBoundRight
+makeEvaluate variables original = do
+    let sortedVariables = sortBy substVariablesFirst variables
```

Comment:
Shouldn't we resimplify the predicate and then resort the variables after substituting the ones that we can? Alternately, we could call `simplifyCondition` before sorting, which may extract some substitutions that are "pendind" inside the predicate, e.g. a postponed map lookup (I don't remember if we still have those).

Reply:
@virgil-serbanuta Does it matter if we call `simplifyCondition` before or after sorting the variables? We don't use the variables when simplifying the condition, and the (newly-formed) condition always gets simplified at the beginning of `makeEvaluateWorker`. Have I misunderstood what you're saying?

Reply:
My question is something like this: can the `simplifyCondition original'` call in `makeEvaluateWorker` add more variables to the substitution? If yes, should that change the sorting order?

As an example, if our `original` condition is `C(x)=f(y) /\ [y=5]`, where `C` is a constructor, `f` is a function and `f(5)=C(A)`, then should we simplify that to `[x=A, y=5]` before sorting, or we should consider `x` as not being a substituted variable?

Now, the above will not be a valid example after https://github.com/kframework/kore/pull/1320 , but you can ask the same about `C(A)=f(x,y) and [y=5]` where `f(x, 5)=C(x)`.

Reply:
Oh, I see. Thank you for the explanation! I'll try to fix this in a follow-up PR.